azure/services/token/client.go

```diff
@@ -0,0 +1,81 @@
+package token
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/policy"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	azureautorest "github.com/Azure/go-autorest/autorest/azure"
+	"github.com/pkg/errors"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+const (
+	defaultEnvironmentName = "AzurePublicCloud"
+)
+
+// AzureClient to get azure active directory token.
+type AzureClient struct {
+	aadToken *azidentity.ClientSecretCredential
+}
+
+// NewClient creates a new azure active directory token client from an authorizer.
+func NewClient(auth azure.Authorizer) (*AzureClient, error) {
+	aadToken, err := newAzureActiveDirectoryTokenClient(auth.TenantID(),
+		auth.ClientID(),
+		auth.ClientSecret(),
+		auth.CloudEnvironment())
+	if err != nil {
+		return nil, err
+	}
+	return &AzureClient{
+		aadToken: aadToken,
+	}, nil
+}
+
+// newAzureActiveDirectoryTokenClient creates a new aad token client from an authorizer.
+func newAzureActiveDirectoryTokenClient(tenantID, clientID, clientSecret, envName string) (*azidentity.ClientSecretCredential, error) {
+	cliOpts, err := getAzureClientOptions(envName)
+	if err != nil {
+		return nil, errors.Wrap(err, "error while getting client options")
+	}
+	clientOptions := &azidentity.ClientSecretCredentialOptions{
+		ClientOptions: cliOpts,
+	}
+	cred, err := azidentity.NewClientSecretCredential(tenantID, clientID, clientSecret, clientOptions)
+	if err != nil {
+		return nil, errors.Wrap(err, "error while getting az client secret credentials")
+	}
+	return cred, nil
+}
+
+func getAzureClientOptions(environment string) (azcore.ClientOptions, error) {
+	if environment == "" {
+		environment = defaultEnvironmentName
```

Comment:
we shouldn't be defaulting the Azure environment inside the clients, we have defaulting in place already in the controller, defaulting in two different places could result in future bugs if the default changes